### PR TITLE
feat(device-client): device-client

### PR DESCRIPTION
**Pull request by ryanbas21**
# JIRA Ticket
https://pingidentity.atlassian.net/jira/software/c/projects/SDKS/boards/1234?selectedIssue=SDKS-3410
# Description
- Still need to implement a better interface for developer usage so they arent calling `store.dispatch` in code like I am in the autoscripts app.
- endpoints are working (can use autoscript file to test, although andy has to readd devices for oath, i accidentally deleted them all)
- need to discuss a testing strategy
- we need to improve types because im not getting them as I expect in the autoscript file. i havent dug into that
# Type of Change

Please Delete options that are not relevant

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Definition of Done

Check all that apply

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Code peer-reviewed.
- [ ] Ensure backward compatibility (special attention).
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] CI build passing on the feature branch.
- [ ] Functional spec is written/updated
- [ ] contains example code snippets.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] UI is completed or ticket is created.
- [ ] Demo to PO and team.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

## Documentation

- [ ] Acceptance criteria met
- [ ] Spell-check run
- [ ] Peer reviewed
- [ ] Proofread

### CODE REVIEW

File: packages/device-client/src/lib/services/index.ts [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        return headers;
+      },
+      baseUrl: Config.get().serverConfig.baseUrl,
+    }),
+    endpoints: (builder) => ({
+      retrieveSessionInfo: builder.mutation<
+        GetSessionInfoResponse,
+        GetSessionInfoQuery | undefined
+      >({
+        async queryFn(query = { realm: '' }, api, _options, baseQuery) {
```
Comment (cerebrl) [inline packages/device-client/src/lib/services/index.ts:35]: If you didn't want to write up the logic for all of these instances of `queryFn`, you could just pass in the realm as a argument, and then just use `query`. That would potentially remove 90% of this code.

File: packages/device-client/src/lib/services/hooks.ts
[the remark this answers is not in the document]
Comment (ryanbas21) [reply packages/device-client/src/lib/services/hooks.ts:1]: The key difference is it uses the /hooks import from react query. they are the same file.

File: packages/device-client/src/lib/services/index.ts [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+        headers.set('Content-Type', 'application/json');
+        headers.set('Accept', 'application/json');
+        headers.set('x-requested-with', 'forgerock-sdk');
+        headers.set('x-requested-platform', 'javascript');
+        return headers;
+      },
+      baseUrl: Config.get().serverConfig.baseUrl,
+    }),
+    endpoints: (builder) => ({
+      retrieveSessionInfo: builder.mutation<
```
[the remark this answers is not in the document]
Comment (ryanbas21) [reply packages/device-client/src/lib/services/index.ts:31]: Screenshot 2024-11-05 at 10 40 17 AM
Yeah i had the same thought, its apparently a post with an empty body. I believe i asked @witrisna also in slack and he said this.

File: packages/device-client/src/lib/services/index.ts (hunk 1 above)
Comment (ryanbas21) [reply packages/device-client/src/lib/services/index.ts:35]: Gonna send msg on slack

File: packages/device-client/src/lib/device-client.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
+import { Config, ConfigOptions } from '@forgerock/javascript-sdk';
+import { configureStore } from '@reduxjs/toolkit';
+import { deviceService } from './services/index.js';
+import { DeleteOathQuery, OathDevice, RetrieveOathQuery } from './types/oath.js';
+import { DeleteDeviceQuery, PushDeviceQuery } from './types/push-device.js';
+import { WebAuthnBody, WebAuthnQuery, WebAuthnQueryWithUUID } from './types/webauthn.js';
+import { BindingDeviceQuery } from './types/binding-device.js';
+
+const deviceStore = () => {
```
[the remark this answers is not in the document]
Comment (ryanbas21) [reply packages/device-client/src/lib/device-client.ts:9]: Can you explain how the factory function helps here? I don't think i follow.

If we have a setConfig at the top level versus in a function, what exactly changes that makes it easier for multiple configurations?

Wouldn't a `config` module be responsible for setting the config and then the consuming package isn't responsible for how to manage the tenant, just reading the configuration information?

I may not fully grasp what the difference is

File: packages/device-client/src/lib/device-client.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(middleware),
+  });
+
+  return {
+    store,
+    endpoints,
+  };
+};
+
+export function setConfig(params: ConfigOptions) {
```
[the remark this answers is not in the document]
Comment (ryanbas21) [reply packages/device-client/src/lib/device-client.ts:27]: see above question

File: packages/device-client/src/lib/services/hooks.ts
[the remark this answers is not in the document]
Comment (ryanbas21) [reply packages/device-client/src/lib/services/hooks.ts:1]: Maybe this is not the best way to do it, i didnt want to spend a lot of time on it, i just copy pasted, used hooks import and renamed the use* functions and exported them.

**Review by cerebrl — APPROVED** [review]

**Pull request merged**